File: easy/Valid Anagram/isAnagram.js
/**
 * @param {string} s
 * @param {string} t
 * @return {boolean}
 */
var isAnagram = function(s, t) {
    const letterMap = {};
    if (s.length !== t.length) {
        return false;
    }

    for (let i = 0; i < s.length; i++) {
        const letter = s[i];
        if (letterMap[letter] != null) {
            letterMap[letter] += 1;
        } else {
            letterMap[letter] = 1;
        }
    }

    for (let i = 0; i < t.length; i++) {
        const letter = t[i];
        if (letterMap[letter] != null) {
            letterMap[letter] -= 1;
            if (letterMap[letter] < 0) {
                return false
            }
        } else {
            return false;
        }
    }

    for (key in Object.keys(letterMap)) {
        if (letterMap[key] > 0) {
            return false;
        }
    }

    return true;
};

const testIsAnagram = function(s, t, output) {
    const result = isAnagram(s, t);
    if (result == output) {
        console.log("Test Case Passed!");
    } else {
        console.log(
          `Test Case Failed: s: ${s} t: ${t} output: ${output} result: ${result}`
        );
    }
}

testIsAnagram("anagram", "nagaram", true);
testIsAnagram("rat", "car", false);
testIsAnagram("", "", true);

File: second_pass/1_arrays_and_hashing/2_valid_anagram/is_anagram.py
class Solution:
    def isAnagram(self, s: str, t: str) -> bool:
        if len(s) != len(t):
            return False

        alphabets = [0] * 26

        offset = ord("a")

        for i in range(len(s)):
            alphabets[ord(s[i]) - offset] += 1
            alphabets[ord(t[i]) - offset] -= 1

        for j in range(26):
            if alphabets[j] != 0:
                return False

        return True

    def testIsAnagram(self, s, t, expected) -> bool:
        actual = self.isAnagram(s, t)
        if actual == expected:
            print("Test Case Passed!")
        else:
            print(f"Test Case Failed! s: {s} t: {t} actual: {actual} expected: {expected}")


solution = Solution()
solution.testIsAnagram("anagram", "nagaram", True)
solution.testIsAnagram("rat", "car", False)
